### PR TITLE
CAP-0010 Fee Bump Account

### PR DESCRIPTION
**Pull request by JeremyRubin**

### CODE REVIEW

File: core/cap-0010.md [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+            ext;
+        } v1;
+    }
+    ext;
+};
+```
+
+When a transaction executes for a given source account, it pays an additional
+fee of `((int64_t) fee_balance) * base_fee` and clears the `fee_balance`. These
+fees are burned and are not inserted into the the Fee Pool. Transactions should not
```
Comment (danrobinson) [inline core/cap-0010.md]: Why?

Comment (danrobinson) [reply core/cap-0010.md]: Ah, I see the discussion below. I feel like this whole mechanism causes a bigger problem than it solves. How about instead we just store the actual amount of the fee_balance that was paid in, and when the extra fee is paid, we just round down to a multiple of `base_fee`? Any remainder could be stuck in the `fee_balance` account, and if you want to get it out you'll have to top up that `fee_balance` to become a multiple of the current base fee.

Comment (JeremyRubin) [reply core/cap-0010.md]: I'm OK with most versions of this, but I want to keep it an uint16 to keep the storage overhead per account low as possible... I'd also like using a 8 bit minifloat and having less fee granularity (e.g., basefee*(1.1**n) where n is 8 bits).

Comment (JeremyRubin) [reply core/cap-0010.md]: Btw, this ends up with the following for a=1.03

This gives you pretty good granularity, and a maximum fee bump of 1877.12 times the base fee.

```haskell
fee :: Uint8 -> Int64
fee n = (table[n]*basefee)/100
where table = 
[100, 103, 106, 109, 112, 115, 119, 122, 126, 130, 134, 138, 142, 146, 151, 155,
160, 165, 170, 175, 180, 186, 191, 197, 203, 209, 215, 222, 228, 235, 242, 250,
257, 265, 273, 281, 289, 298, 307, 316, 326, 335, 346, 356, 367, 378, 389, 401,
413, 425, 438, 451, 465, 479, 493, 508, 523, 539, 555, 572, 589, 606, 625, 643, 
663, 682, 703, 724, 746, 768, 791, 815, 840, 865, 891, 917, 945, 973, 1003,
 1033, 1064, 1096, 1128, 1162, 1197, 1233, 1270, 1308, 1347, 1388, 1430,
 1472, 1517, 1562, 1609, 1657, 1707, 1758, 1811, 1865, 1921, 1979, 2038,
 2100, 2163, 2227, 2294, 2363, 2434, 2507, 2582, 2660, 2740, 2822, 2906,
 2994, 3084, 3176, 3271, 3369, 3471, 3575, 3682, 3792, 3906, 4023, 4144
, 4269, 4397, 4528, 4664, 4804, 4948, 5097, 5250, 5407, 5570, 5737, 5909,
 6086, 6269, 6457, 6650, 6850, 7056, 7267, 7485, 7710, 7941, 8179, 8425,
 8678, 8938, 9206, 9482, 9767, 10060, 10362, 10672, 10993, 11322, 11662, 
12012, 12372, 12743, 13126, 13520, 13925, 14343, 14773, 15216, 15673, 
16143, 16627, 17126, 17640, 18169, 18714, 19276, 19854, 20450, 21063,
 21695, 22346, 23017, 23707, 24418, 25151, 25905, 26683, 27483, 28308, 
29157, 30032, 30932, 31860, 32816, 33801, 34815, 35859, 36935, 38043,
 39184, 40360, 41571, 42818, 44103, 45426, 46788, 48192, 49638, 51127, 
52661, 54241, 55868, 57544, 59270, 61048, 62880, 64766, 66709, 68711, 
70772, 72895, 75082, 77334, 79654, 82044, 84505, 87041, 89652, 92341,
 95112, 97965, 100904, 103931, 107049, 110261, 113568, 116975, 120485, 
124099, 127822, 131657, 135607, 139675, 143865, 148181, 152627, 157205, 
161922, 166779, 171783, 176936, 182244, 187712]

Comment (danrobinson) [reply core/cap-0010.md]: > I'm OK with most versions of this, but I want to keep it an uint16 to keep the storage overhead per account low as possible... I'd also like using a 8 bit minifloat and having less fee granularity (e.g., basefee*(1.1**n) where n is 8 bits).

That strikes me as pathological overoptimization.

Comment (JeremyRubin) [reply core/cap-0010.md]: @MonsieurNicolas and @jonjove seem to be pretty paranoid about account entry size growth, ask them.

File: core/cap-0010.md [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+            Liabilities liabilities;
+
+            union switch (int v)
+            {
+            case 0:
+                void;
+			case 1:
+				struct
+				{
+					uint16 fee_balance;
```
Comment (bartekn) [inline core/cap-0010.md]: This name is confusing. What do you think about `base_fee_multiplier`?

**Pull request merged**